controller/kubernetes_configmap_controller.go

```diff
@@ -0,0 +1,246 @@
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(logrus.Infof)
+	// TODO: remove the wrapper when every clients have moved to use the clientset.
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{
+		Interface: v1core.New(kubeClient.CoreV1().RESTClient()).Events(""),
+	})
+
+	kc := &KubernetesConfigMapController{
+		baseController: newBaseController("longhorn-kubernetes-configmap-controller", logger),
+
+		namespace:    namespace,
+		controllerID: controllerID,
+
+		ds: ds,
+
+		kubeClient:    kubeClient,
+		eventRecorder: eventBroadcaster.NewRecorder(scheme, v1.EventSource{Component: "longhorn-kubernetes-configmap-controller"}),
+
+		cfmStoreSynced: configMapInformer.Informer().HasSynced,
+	}
+
+	configMapInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    kc.enqueueConfigMapChange,
+		UpdateFunc: func(old, cur interface{}) { kc.enqueueConfigMapChange(cur) },
+		DeleteFunc: kc.enqueueConfigMapChange,
+	})
+
+	return kc
+}
+
+func (kc *KubernetesConfigMapController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer kc.queue.ShutDown()
+
+	kc.logger.Infof("Start")
+	defer kc.logger.Infof("Shutting down")
+
+	if !controller.WaitForCacheSync(kc.name, stopCh, kc.cfmStoreSynced) {
+		return
+	}
+	for i := 0; i < workers; i++ {
+		go wait.Until(kc.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (kc *KubernetesConfigMapController) worker() {
+	for kc.processNextWorkItem() {
+	}
+}
+
+func (kc *KubernetesConfigMapController) processNextWorkItem() bool {
+	key, quit := kc.queue.Get()
+	if quit {
+		return false
+	}
+	defer kc.queue.Done(key)
+	err := kc.syncHandler(key.(string))
+	kc.handleErr(err, key)
+	return true
+}
+
+func (kc *KubernetesConfigMapController) handleErr(err error, key interface{}) {
+	if err == nil {
+		kc.queue.Forget(key)
+		return
+	}
+
+	if kc.queue.NumRequeues(key) < maxRetries {
+		kc.logger.WithError(err).Warnf("Error syncing configmap %v", key)
+		kc.queue.AddRateLimited(key)
+		return
+	}
+
+	kc.logger.WithError(err).Warnf("Dropping configmap %v out of the queue", key)
+	kc.queue.Forget(key)
+	utilruntime.HandleError(err)
+}
+
+func (kc *KubernetesConfigMapController) syncHandler(key string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "%v: fail to sync %v", kc.name, key)
+	}()
+
+	namespace, cfmName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	if err := kc.reconcileDefaultStorageClass(namespace, cfmName); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (kc *KubernetesConfigMapController) reconcileDefaultStorageClass(namespace, cfmName string) error {
+	// The default storageclass configmap must be in the same namespace as the current KubernetesConfigMapController
+	if namespace != kc.namespace {
+		return nil
+	}
+
+	storageCFM, err := kc.ds.GetConfigMap(kc.namespace, types.DefaultStorageClassConfigMapName)
```

Comment:
If the configmap is not the one we want, we don't need to go through this.

There can be other config maps.

Reply:
Thank you. Fixed